# HACS202 Final Project
## Data Analysis Section
## Group A, Spring 2020

Akilesh Praveen

This notebook showcases the scraping of data from honeypot logs, the tidying of said data, and the interpretation of this data.

This repository also contains the raw scripts used to perform data scraping, with additional debug features included. Find them in the 'scripts' directory.

### Section 0: Imports

Libraries and initialization for this project.

In [1]:
import re
import pandas as pd

### Section 1: Data Collection

First, let's extract the data from the `mitm_files` folder. This script is a little involved, but it essentially separates the log into 'attacks'. Each attack is considered begun when the script detects a connection initiated, and each attack is considered ended when the script detecs that the connection is closed. From this, we extract the date & time of the attack, whether a password was used, the IP address of the attacker, and the client that the attacker was using.

This data was then transferred from a list of lists to a dataframe.

_When connections were established and then not closed, that data appeared as ambiguous, and was therefore excluded._

In [2]:
def analyze_attack(fp):
    line = fp.readline()

    if not line:
        return None

    # read the first line of an attack
    initial = re.search("([0-9]{4}-[0-9]{2}-[0-9]{2})\s([0-9]{2}:[0-9]{2}:[0-9]{2})\.[0-9]{3}\s-\s\[Debug\]\s\[Connection\]\sAttacker\sconnected:\s(.*)\s\|\sClient\sIdentification:\s(.*)", line)

    if initial:
        my_date = initial.group(1)
        my_time = initial.group(2)
        my_ip = initial.group(3)
        my_client = initial.group(4)

        # number of attempts + whether a password was used or not
        my_attempts = 0
        my_passwords = []

        # this attack continues until we find the 'Attacker closed the connection' line
        end = False
        while(not end):
            nextline = fp.readline()
            found_end = re.search("Attacker\sclosed\sthe\sconnection$",nextline)

            if found_end:
                end = True
                return [my_date, my_time, my_ip, my_client, my_attempts, my_passwords]

            # this line is not the end, we can get data from it
            
            # check to see if password attempts have increased
            attempt_add = re.search("has\sso\sfar\smade", nextline)

            if attempt_add:
                my_attempts += 1

            password_add = re.search("trying\sto\sauthenticate\swith\s\"(.*)\"", nextline)

            if password_add:
                if password_add.group(1) == "none":
                    my_passwords.append(False)
                else:
                    my_passwords.append(True)

            




    else:
        # not a distinct 'attack'- move to next event
        read_until_next(fp)
        return ["", "", "", "", "", []]


def read_until_next(fp):
    line = fp.readline()

    while (line):
        nextgroup = re.search("Attacker\sclosed\sthe\sconnection$",line)

        if nextgroup:
            break
        else:
            line = fp.readline()
            
# use analyze_attack to store all attacks in a list for now


mitm_data_102 = []
mitm_data_103 = []

# open all 3 files and analyze them

with open('../raw_data/mitm_files/mitm_file102', 'r') as fp:
    # fp is now our file pointer to the mitm data file.

    # skip the five lines at the top
    line = fp.readline()
    line = fp.readline()
    line = fp.readline()
    line = fp.readline()
    line = fp.readline()

        
    while line:
        curr = analyze_attack(fp)
        if curr is None:
            break
        else:
            mitm_data_102.append(curr)


            
# move the contents of data to a dataframe
mitm_df_102 = pd.DataFrame(mitm_data_102, columns = ['date', 'time', 'ip_address', 'client_type', 'attempts', 'password_used' ])   
mitm_df_102

date      time      ip_address  \
0      2020-05-04  15:14:00    195.3.147.47   
1      2020-05-04  15:14:08    195.3.147.47   
2      2020-05-04  15:14:20    195.3.147.47   
3      2020-05-04  15:16:08  193.105.134.45   
4      2020-05-04  15:16:19   103.89.89.242   
5                                             
6      2020-05-04  15:19:23  193.105.134.45   
7      2020-05-04  15:22:59    195.3.147.47   
8      2020-05-04  15:24:47    195.3.147.47   
9      2020-05-04  15:27:16    195.3.147.47   
10     2020-05-04  15:27:34    195.3.147.47   
11     2020-05-04  15:28:18  193.105.134.45   
12     2020-05-04  15:31:25    195.3.147.47   
13     2020-05-04  15:37:25    195.3.147.47   
14     2020-05-04  15:38:25    195.3.147.47   
15     2020-05-04  15:40:22  193.105.134.45   
16     2020-05-04  15:43:05    195.3.147.47   
17     2020-05-04  15:44:13   109.236.91.85   
18     2020-05-04  15:44:40    195.3.147.47   
19     2020-05-04  15:48:33  193.105.134.45   
20                                            
21     2020-05-04  15:53:29  193.105.134.45   
22     2020-05-04  15:56:31    195.3.147.47   
23     2020-05-04  15:56:41  193.105.134.45   
24     2020-05-04  15:57:53    195.3.147.47   
25     2020-05-04  16:02:04    195.3.147.47   
26     2020-05-04  16:03:16    195.3.147.47   
27     2020-05-04  16:03:57  193.105.134.45   
28     2020-05-04  16:04:00   109.236.91.85   
29     2020-05-04  16:11:05   109.236.91.85   
...           ...       ...             ...   
18876  2020-05-11  02:33:34    195.3.147.47   
18877  2020-05-11  02:35:28    195.3.147.47   
18878  2020-05-11  02:35:59  193.105.134.45   
18879  2020-05-11  02:40:58  193.105.134.45   
18880  2020-05-11  02:42:28  193.105.134.45   
18881  2020-05-11  02:47:40  193.105.134.45   
18882  2020-05-11  02:51:26   109.236.91.85   
18883  2020-05-11  02:53:25  193.105.134.45   
18884  2020-05-11  02:56:50    195.3.147.47   
18885  2020-05-11  02:59:11    195.3.147.47   
18886  2020-05-11  03:00:56    195.3.147.47   
18887                                         
18888  2020-05-11  03:05:48   14.248.209.96   
18889  2020-05-11  03:10:07  193.105.134.45   
18890  2020-05-11  03:10:19   109.236.91.85   
18891  2020-05-11  03:10:46   109.236.91.85   
18892  2020-05-11  03:15:36  193.105.134.45   
18893  2020-05-11  03:17:38    195.3.147.47   
18894  2020-05-11  03:17:50    195.3.147.47   
18895  2020-05-11  03:19:20  193.105.134.45   
18896  2020-05-11  03:21:13  193.105.134.45   
18897  2020-05-11  03:25:31    195.3.147.47   
18898  2020-05-11  03:26:56    195.3.147.47   
18899  2020-05-11  03:27:52  193.105.134.45   
18900  2020-05-11  03:28:04  193.105.134.45   
18901  2020-05-11  03:31:31  109.236.81.198   
18902  2020-05-11  03:32:36    195.3.147.47   
18903                                         
18904  2020-05-11  03:38:37    195.3.147.47   
18905  2020-05-11  03:38:51  180.183.248.39   

                             client_type attempts               password_used  
0                 SSH-2.0-paramiko_1.8.1        1               [False, True]  
1                    SSH-2.0-OpenSSH_5.3        1               [False, True]  
2           SSH-2.0-WinSCP_release_5.1.5        1               [False, True]  
3           SSH-2.0-WinSCP_release_5.7.5        1               [False, True]  
4                 SSH-2.0-paramiko_2.7.1        2         [True, False, True]  
5                                                                          []  
6                    SSH-2.0-OpenSSH_5.2        1               [False, True]  
7           SSH-2.0-WinSCP_release_5.1.5        1               [False, True]  
8                  SSH-2.0-OpenSSH_3.9p1        1               [False, True]  
9                 SSH-2.0-paramiko_1.8.1        1               [False, True]  
10               SSH-2.0-paramiko_1.15.1        1               [False, True]  
11                   SSH-2.0-OpenSSH_4.3        1               [False, True]  
12                   SSH-2.0-OpenSSH_6.1        1         

In [3]:
# make a similar dataframe for log 103

with open('../raw_data/mitm_files/mitm_file103', 'r') as fp:
    # fp is now our file pointer to the mitm data file.

    # skip the five lines at the top
    line = fp.readline()
    line = fp.readline()
    line = fp.readline()
    line = fp.readline()
    line = fp.readline()

        
    while line:
        curr = analyze_attack(fp)
        if curr is None:
            break
        else:
            mitm_data_103.append(curr)
            
mitm_df_103 = pd.DataFrame(mitm_data_103, columns = ['date', 'time', 'ip_address', 'client_type', 'attempts', 'password_used' ])   
mitm_df_103

date      time      ip_address                   client_type  \
0      2020-05-04  15:29:40  193.105.134.45          SSH-2.0-Granados-1.0   
1      2020-05-04  15:45:47    195.3.147.47      SSH-2.0-paramiko_1.7.7.1   
2      2020-05-04  15:56:02  180.214.238.55                    SSH-2.0-Go   
3      2020-05-04  16:01:08    195.3.147.47           SSH-2.0-OpenSSH_5.2   
4      2020-05-04  16:08:14  193.105.134.45          SSH-2.0-libssh_0.5.5   
5      2020-05-04  16:31:13    195.3.147.47           SSH-2.0-OpenSSH_6.2   
6      2020-05-04  16:45:28    195.3.147.47          SSH-2.0-libssh_0.5.5   
7      2020-05-04  16:53:34    195.3.147.47           SSH-2.0-OpenSSH_5.9   
8      2020-05-04  16:55:27  193.105.134.45    SSH-2.0-PuTTY_Release_0.64   
9      2020-05-04  17:12:14    195.3.147.47   SSH-2.0-Nmap_SSH2_Hostkey     
10     2020-05-04  17:13:13  193.105.134.45  SSH-2.0-WinSCP_release_4.1.9   
11     2020-05-04  17:42:53    195.3.147.47          SSH-2.0-libssh_0.5.5   
12     2020-05-04  17:59:36    195.3.147.47    SSH-2.0-PuTTY_Release_0.64   
13     2020-05-04  18:03:32   109.236.91.85        SSH-2.0-paramiko_1.7.5   
14     2020-05-04  18:06:09  193.105.134.45       SSH-2.0-paramiko_1.16.0   
15     2020-05-04  18:16:31  193.105.134.45          SSH-2.0-libssh_0.5.5   
16     2020-05-04  18:36:22    195.3.147.47           SSH-2.0-OpenSSH_6.0   
17     2020-05-04  18:41:08    195.3.147.47        SSH-2.0-paramiko_2.0.2   
18     2020-05-04  18:46:45    141.98.81.83           SSH-2.0-OpenSSH_7.3   
19     2020-05-04  18:46:47    141.98.81.84           SSH-2.0-OpenSSH_7.3   
20     2020-05-04  18:46:52    141.98.81.99           SSH-2.0-OpenSSH_7.3   
21     2020-05-04  18:46:56   141.98.81.107           SSH-2.0-OpenSSH_7.3   
22     2020-05-04  18:47:00   141.98.81.108           SSH-2.0-OpenSSH_7.3   
23     2020-05-04  18:47:04    141.98.81.81           SSH-2.0-OpenSSH_7.3   
24     2020-05-04  18:47:08    141.98.81.83           SSH-2.0-OpenSSH_7.3   
25     2020-05-04  18:47:12    141.98.81.84           SSH-2.0-OpenSSH_7.3   
26     2020-05-04  18:47:15    141.98.81.99           SSH-2.0-OpenSSH_7.3   
27     2020-05-04  18:47:18   141.98.81.107           SSH-2.0-OpenSSH_7.3   
28     2020-05-04  18:47:22   141.98.81.108           SSH-2.0-OpenSSH_7.3   
29     2020-05-04  18:47:26    141.98.81.81           SSH-2.0-OpenSSH_7.3   
...           ...       ...             ...                           ...   
15705  2020-05-11  02:42:54    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15706  2020-05-11  02:43:09    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15707  2020-05-11  02:43:24    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15708  2020-05-11  02:43:36    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15709  2020-05-11  02:43:48    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15710  2020-05-11  02:44:00    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15711  2020-05-11  02:44:13    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15712  2020-05-11  02:44:24    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15713  2020-05-11  02:44:36    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15714  2020-05-11  02:44:47    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15715  2020-05-11  02:44:59    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15716  2020-05-11  02:45:11    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15717  2020-05-11  02:45:22    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15718  2020-05-11  02:45:34    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15719  2020-05-11  02:45:45    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15720  2020-05-11  02:45:57    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15721  2020-05-11  02:46:08    5.101.151.82         SSH-2.0-libssh2_1.4.3   
15722  2020-05-11  02:51:31  103.145.253.87        SSH-2.0-paramiko_2.7.1   
15723  2020-05-11  02:53:55  193.105.134.45           SSH-2.0-PuTTY_KiTTY   
15724  2020-05-11  03:11:19    195.3.147.47           SSH-2.0-OpenSSH_6.0   
15725  20

#### Container 104

Interesitngly enough, container 104 had such an erratic Connect and Disconnect pattern that separate events were nigh-indistinguishable. Data from it was not easily convertible into a format similar to containers 102 and 103.

Although not quantitatively definable at this time, this is an interesting observation to make regardless. We can already see this as a clear difference between containers 102, 103, and 104.

#### Login attempts

We will now look to collect the login attempts for all of our containers, and their details. This can be found in the `raw_data/logins` folder.

Like before, we'll read the data into lists of lists, then proceed to represent them as dataframes.

### Section 2: Tidying Data

Using Python and Pandas features to tidy the data we have placed in dataframes.

### Section 3: Data Analysis

Drawing conclusions from the data we've come up with.